src/library/scala/collection/Seq.scala

```diff
@@ -857,12 +857,16 @@ trait SeqOps[+A, +CC[_], +C] extends Any
   def diff[B >: A](that: Seq[B]): C = {
     val occ = occCounts(that)
     fromSpecific(iterator.filter { x =>
-      val ox = occ(x)  // Avoid multiple map lookups
-      if (ox == 0) true
-      else {
-        occ(x) = ox - 1
-        false
+      var include = false
+      occ.updateWith(x) {
+        case None => {
+          include = true
+          None
+        }
+        case Some(1) => None
+        case Some(n) => Some(n - 1)
```

Comment:
So are these implementations trading off faster size=0/size=1 operations at the cost of these Some allocations?  Because that doesn't sound performant.  So perhaps I'm misunderstand, but I thought I would ask.

Reply:
This should involve fewer allocations and fewer lookups in most cases.

In this PR, updateWith to decrement (n >1) costs a hash lookup (to find the bucket), a Some allocation (to return the result) and another Some allocation (to hold de decremented result).

In the situation before, there was a hash lookup (to find the bucket), a Some allocation (to get the value to apply), then another hash lookup to find the bucket again, and then another default map allocation.

So the decrement costs an extra Some allocation, but saves a WithDefaultValue allocation *and* a hash lookup.

The other cases only look better: when the returned value is 1, now we don't have the  extra Some allocation, and we clear the bucket, and maybe thay empties the map an get to structurally share the rest, we short circuit and append an structurally share the tail.

The last case is if the element is not in the underlying map. In the old case, that costs just the lookup, and a Some allocation iff there was something in the map before, and if not, just the lookup. In the new situation, it also costs just the lookup.

The 0 and 1 special casing is only at the start, where this avoids building the map entirely and could be done independently of the code in this part.